### ISSUE TITLE
Add equality comparer to foreach binding

### ISSUE THREAD

**Issue by bennieswart** [knockout/knockout#2298]
Currently the `foreach` binding compares array elements for equality using `===` always. This is fine in most cases where the array elements are indeed the same.

In one of our use cases, we have a diary with a list of bookings. An algorithm runs on the bookings to determine their position and size on the screen, and for each booking, it returns an object wrapping the booking, like
```js
{
  booking: thebooking,
  render: {
    top: 0,
    height: 0.05,
    left: 0,
    width: 1,
    ...
  }
};
```
The array of these wrapper objects are then passed to `foreach` for rendering. As you might expect, any change results in a redraw of all bookings, since the wrapper objects are never the exact same objects (by reference), although they look identical.

Workarounds to this problem might include passing the array of bookings to `foreach`, and storing the render parameters on the booking instance. This is undesirable as it stores unnecessary information on the booking instance.
Another workaround might be to maintain a parallel array of render parameters in addition to the array of bookings. This has some extra work associated with it, because in order to detect changes in individual parameters, they would have to be made observable, and this would require careful updating of the render parameters array instead of simply generating a new one.

I suggest adding an additional option to the `foreach` (and `template`) bindings, something to the effect of `equals` or `equalityComparer`, which takes a function taking two parameters and returning `true`/`false` if they are equal/not equal. The default value for this option would be `function(a,b){return a===b}`, such that, by default, everything continues working as it currently does.

In the case of the bookings, one could then specify `function(a,b){return a.booking===b.booking&&JSON.stringify(a.render)===JSON.stringify(b.render)}`. Similar solutions would be possible in like cases.

The patch is only a handful of lines and I would be happy to submit a PR if this suggestion gains approval.

**Comment by mbest** [comment]
This would not really work the way you think it would. It would mean that array items that are "unchanged" would still be bound to the old objects rather than the new ones. This would be a real problem if you change observables in the new object, since Knockout wouldn't see the change.

**Comment by bennieswart** [comment]
I understand that part and I agree, but that is exactly why the comparitor function would have to include those fields in its comparison, as demonstrated rather crudely with the `JSON.stringify`. Obviously you can do a more specific comparison than the stringify.

In the above example, if the comparitor did *not* include the `render` fields in its comparison, and those render fields were bound in the ui, then one would en up with the problem you are describing.

Furthermore, imagine a case where additional properties are present on the array item, which are used elsewhere in the application, but not in that specific `foreach`. The comparitor function would allow you to explicitly ignore any changes in those properties (since they would not change anything on the ui).

In summary:
Currently, if the exact value of the array item changes (`a === b`), a redraw is performed.
Having an explicit comparitor function (`equals(a, b)`), would reduce the number of redraws.
This is a problem if the redraws are reduced 'too much' such that changes in fields bound to the ui are not detected (your scenario).
It is not a problem when the comparitor function takes into account all fields bound to the ui. This offers significant performance increase and flexibility with the type of objects that can be used in the loop.
